Add a variant of Wasm image spec based on standard media types.

mathetake opened the pull request:
**Description**

Following the discussion (led by @dgn) in the Istio Wasm-SIG, this PR added a variant of Wasm image spec for storing Wasm binaries as standard container images. 

Notably this new spec leverages the existing standard layers with either one of
- `application/vnd.oci.image.layer.v1.tar+gzip` or 
- `application/vnd.docker.image.rootfs.diff.tar.gzip`

media type, not based on the OCI Artifact Spec. This allows users to use standard CLI tools in container ecosystem (such as `docker`, `podman`, `buildah`, etc.) and push to any standard registries without requiring support for arbitrary mediaType (including the one used before).

See #208 and [a meeting note by @dgn](https://docs.google.com/document/d/1Uf-3RPFRyNl6ExYUeZERlB9Y8kyxZFaB6NbDZyguXwI/edit) for the background, and appendix sections here for how end-users would operate to build images.


cc @dgn, @yuval-k @PiotrSikora @bianpengyuan @kyessenov 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

resolves #208

dgn commented:
hi @mathetake! Thanks for working on this.

I believe there has been a misunderstanding. The idea with the OCI-compliant spec, as agreed on with @yuval-k , was not to replace the existing spec, but to add a _variant_ that would only use OCI-compliant mediaTypes. Clients would then be able to look at the manifest and determine (e.g. from an annotation on the manifest) which variant is being used and parse it accordingly - the main difference being that the OCI-compliant variant uses a tarball to package the files, whereas the existing format uses no compression and custom mediaTypes.

That also means we'll want to keep everything that is already in the spec (e.g. all the envoy-specific configuration) and make it available in both variants. We don't want to break any existing users

mathetake commented:
@dgn ah right, thanks for clarifying.. will fix accordingly. Thanks.

mathetake commented:
@dgn fixed. PTAL.

dgn commented on spec/README-compat.md:
````diff
@@ -0,0 +1,84 @@
+Then copy the files into that base image by `buildah copy` command to create the layer.
+
+```
+$ buildah copy mywasm filter.wasm runtime-config.json ./
````
this probably also needs a `buildah config --annotation module.wasm.image/variant=compat mywasm` (I'm currently unable to test whether that actually produces the manifest shown above)

mathetake replied:
I thought I added that instruction, but yeah thanks. will fix.

mathetake replied:
fixed.

dgn commented on spec/README-compat.md:
```diff
@@ -0,0 +1,84 @@
+
+## Description
+
+This *compat* variant makes use of `application/vnd.oci.image.layer.v1.tar+gzip` media type for layers, and is not based on custom OCI Artifcat media types. This way users can oeperate with standard tools such as docker, podman, buildah, etc.
```
we should probably also allow the docker mediaType `application/vnd.docker.image.rootfs.diff.tar.gzip` - it's the same thing anyway, and it opens us up to supporting image building using `docker build`

mathetake replied:
yeah good point. Let me check how we could embed annotation in the docker format first.

mathetake replied:
OK we can use `"mediaType": "application/vnd.docker.distribution.manifest.v2+json"` in the docker manifest without using additional docker command. cool.

mathetake replied:
implemented the here https://github.com/istio/istio/compare/master...mathetake:wasm-image-fetch?expand=1

mathetake replied:
Fixed in https://github.com/solo-io/wasm/pull/265/commits/93e2c06c14be4fd4b91be98bbef39c3eac199437.

mathetake commented:
@dgn PTAL again when you are back! 🙇  [Here's](https://github.com/istio/istio/compare/master...mathetake:wasm-image-fetch?expand=1) the fetcher implementation for both docker and oci compat images, which I open a PR in Istio once this lands.

mathetake commented:
cc @mandarjog (I forgot to cc you).

mathetake commented:
JFYI @lizan

mandarjog commented on spec/README-compat.md:
```diff
@@ -0,0 +1,139 @@
+# Wasm Image Specification v0.0.0
+
+## Introduction:
+
```
Can we add some text to the main spec about the availability of a compat format?

mathetake replied:
OK agreed. will do.

mathetake replied:
intentionally not to apply your suggestion in favor of added general info in README.md. PTAL

mandarjog commented on spec/README-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+In addition, the layer must consist of the following two files:
+- `plugin.wasm` - (**Required**) A Wasm binary to be loaded by the runtime.
+- `runtime-config.json` - (**Optional**) A runtime configuratio specified in [Wasm Artifact Image Specification](README.md).
```
Link to specific anchor is preferred showing runtime-config.json.

mandarjog replied:
Somewhere is should be clear that plugin.wasm and a json file is the common payload, and we have 3 ways of packaging them based on your needs.

mathetake replied:
yeah agreed. Maybe better add a `DEVELOPER.md` or any separate markdown for explaining these three ways and how Istio will support them. sg?

mathetake replied:
done. https://github.com/solo-io/wasm/pull/265/commits/ee52cb636f7f2fbf4054a8885efcac690def0b55

mathetake commented on spec/README-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+## Description
+
+This *compat* variant makes use of standard media type for layers, and is not based on custom OCI Artifcat media types. This way users can oeperate with standard tools such as docker, podman, buildah, etc.
```
Thank you! 😄

mathetake commented on spec/README-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+We demonstrate how to build a *compat* image with Docker CLI. Produced images have `application/vnd.docker.image.rootfs.diff.tar.gzip` layer media type.
+
+We assume that you have a valid Wasm binary named `filter.wasm` and `runtime-config.json` that you want to package as an image.
```
yeah, thanks!

mathetake commented:
restructured the spec/ directory and added a general instruction as suggested by @mandarjog. PTAL folks https://github.com/mathetake/wasm/tree/oci-compliant-spec/spec

mathetake commented on spec/spec.md:
```diff
@@ -0,0 +1,133 @@
+
```
note: this is moved from the former README.md as-is.

Sodman replied:
I commented above too but I'm thinking we should move the following sections (remove from here and keep at the top level spec/README.md), as they're relevant to all variants, not just the *oci* variant:

- Introduction
- Terminology
- Overview section from within Description

mathetake replied:
OK I bring the common terminology out of spec.md to README.md, and left the spec.md specific ones as-is.

mathetake commented:
https://github.com/istio/istio/compare/master...mathetake:wasm-image-fetch?expand=1 and there's already working fetcher implementation in Istio: https://github.com/istio/istio/compare/master...mathetake:wasm-image-fetch?expand=1

Sodman reviewed (commented):
Thanks again for working on this! I left a few comments, mostly about layout. I think we should keep the high-level overview from the original *oci* variant at the top level readme, and then have each of the variant spec readmes simply document their specific implementations, layers, etc.

mathetake commented on spec/spec-compat.md:
````diff
@@ -0,0 +1,139 @@
+Then copy the files into that base image by `buildah copy` command to create the layer.
+
+```
+$ buildah copy mywasm filter.wasm runtime-config.json ./
````
https://github.com/solo-io/wasm/pull/265/commits/8098c502d53c91bad25d68d57b52912e3c8c952b done, thanks!

mathetake commented on spec/spec-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+We demonstrate how to build a *compat* image with Buildah, a standard cli for building OCI images. We use v1.21.0 of Buildah here. Produced images have `application/vnd.oci.image.layer.v1.tar+gzip` layer media type
+
+We assume that you have a valid Wasm binary named `filter.wasm` and `runtime-config.json` that you want to package as an image.
```
https://github.com/solo-io/wasm/pull/265/commits/8098c502d53c91bad25d68d57b52912e3c8c952b done, thanks!

mathetake commented on spec/README.md:
```diff
@@ -1,133 +1,49 @@
-- [Envoy WASM Filter Specification](#envoy-wasm-filter-specification)
-    - [Example](#example)
+Here we have several specifications for how to bundle Wasm modules as container images. 
+They primarily consider the use case of [Envoy] Wasm plugins in [Istio] service mesh.
```
https://github.com/solo-io/wasm/pull/265/commits/bff526ce591216a4de7d014c3e17ed562bb04ef0 done!

mathetake commented on spec/spec.md:
```diff
@@ -0,0 +1,114 @@
+
+| Term                               | Definition                                       |
+|------------------------------------|--------------------------------------------------|
+| Envoy WASM Filter                  | Custom Filters for the Envoy Proxy built as a WASM module.
+| Envoy WASM OCI Image               | Envoy Filters stored as OCI images according to the specification. 
+| Envoy WASM OCI Artifact Specification | An extension of the WASM OCI Artifact Spec which describes how to bundle and ship Envoy WASM filters as OCI Images. |
```
note: these are *oci* variant specific terms only used in spec.md

mathetake commented on spec/README.md:
```diff
@@ -1,133 +1,61 @@
+| Wasm Module                        | The distributable, loadable, and executable unit of code in WebAssembly. 
+| Wasm Image Specification           | The specification for storing Wasm modules as container images.
+| Wasm Runtime                       | The execution environment into which a Wasm Module may be loaded. This refers to the application itself which loads and executes a wasm module. Examples include web browsers, the Open Policy Agent, the Envoy Proxy, or any other application which supports extension via Wasm modules. 
+| Rurntime Configuation              | Configuration or metadata specific to the runtime which consumes a module. 
```
Thanks! :)

mathetake commented:
@dgn @mandarjog PTAL!

dgn commented on spec/spec-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+## Introduction:
+
+This document describes a varient of [Wasm Artifact Image Specification](spec.md), which leverages the compatible layer media types. Here, we omit definition and terminology explained in [Wasm Artifact Image Specification](spec.md). 
```
typo: `variant`

dgn commented on spec/spec-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+## Description
+
+This *compat* variant makes use of compatible media type for layers, and is not based on custom OCI Artifact media types. This way users can oeperate with standard tools such as docker, podman, buildah, etc.
```
typo: `operate`

maybe also mention that this is strictly for compatibility, ie for registries that don't yet support custom mediaTypes in OCI images, or for existing toolchains that require standard docker images

mathetake replied:
done! https://github.com/solo-io/wasm/pull/265/commits/d06e5880695665da98d90f770a9c741ba94ba2fa

dgn commented on spec/spec-compat.md:
```diff
@@ -0,0 +1,139 @@
+
+### Annotation
+
+If the media type equals `application/vnd.oci.image.layer.v1.tar+gzip`, then a *compat* variant image must add the annotation `module.wasm.image/variant=compat` in the manifest.
```
Why only for this media type? If it is not easily added by the docker cli, maybe we can just relax this requirement to a "should"

mathetake replied:
OK I changed the implementation so that we remove the need for annotations: https://github.com/istio/istio/commit/9975a786bb104088f5561223c320f8b507378ffa

Will change the text.

mathetake replied:
https://github.com/solo-io/wasm/pull/265/commits/45316b2bdafe98962c87ac675f46499f59722d16 done.

dgn commented on spec/README.md:
```diff
@@ -1,133 +1,61 @@
-|------------|------|-------------|
-| application/vnd.module.wasm.config.v1+json | JSON Object | Configuration for the Target WASM runtime.
-| application/vnd.module.wasm.content.layer.v1+wasm | binary data (byte array) | The compiled module data |
+The key difference between these two variants is that, the *oci* variant makes use of the custom media types on [OCI Artifact] for image layers while the *compat* variant leverages the standard media types.
```
Maybe we should add a sentence like "our goal is to make the OCI variant the default format for shipping Wasm modules in container images, we acknowledge however that there are toolchains and registries deployed and in use that do not support our custom media types yet. To accomodate those exisiting toolchains, there is the semantically equivalent *compat* variant, which provides the same feature set, but is compatible with existing tooling because it 'looks' very much like a normal container image. Implementations of this spec should support both variants."

mathetake replied:
done! https://github.com/solo-io/wasm/pull/265/commits/d06e5880695665da98d90f770a9c741ba94ba2fa

dgn reviewed (approved):
lgtm so far, a few comments

mathetake commented:
Thanks @dgn, addressed all your comments and changed the implementation accordingly.

@Sodman @mandarjog I believe this is ready to land. I already have the implementation here (https://github.com/istio/istio/compare/master...mathetake:wasm-image-fetch?expand=1) which supports all the variants 
 in Istio. I will open the PR against Istio once this lands.

Sodman reviewed (approved):
LGTM 👍 Thanks for putting this together!

jonjohnsonjr commented on spec/spec.md:
```diff
@@ -0,0 +1,114 @@
+    }
+  },
+  {
+    "mediaType": "application/vnd.module.wasm.content.layer.v1+wasm",
```
This is still an odd choice for a mediaType, see https://github.com/solo-io/wasm/issues/256

mathetake replied:
Yes this is just moved from the original one as-is. See the description and comments in the PR.

mathetake commented:
fyi I just opened a PR for fetching all the variants in Istio: https://github.com/istio/istio/pull/33812

jcchavezs commented on spec/spec-compat.md:
````diff
@@ -0,0 +1,142 @@
+COPY runtime-config.json plugin.wasm ./
+```
+
+**Note: you must have exactly one `COPY` instruction in the Dockerfile in order to end up having only one layer in produced images**
````
This kind of sounds confusing to me, is it **exactly one COPY layer** or **exactly ONE layer**?
